helix-view/src/document.rs

```diff
@@ -407,7 +409,57 @@ impl Document {
 
     /// If supported, returns the changes that should be applied to this document in order
     /// to format it nicely.
-    pub fn format(&self) -> Option<impl Future<Output = LspFormatting> + 'static> {
+    // We can't use anyhow::Result here since the output of the future has to be
+    // clonable to be used as shared future. So use a custom error type.
+    pub fn format(&self) -> Option<BoxFuture<'static, Result<Transaction, FormatterError>>> {
+        if let Some(formatter) = self.language_config().and_then(|c| c.formatter.clone()) {
+            use std::process::Stdio;
+            let text = self.text().clone();
+            let mut process = tokio::process::Command::new(&formatter.command);
+            process
+                .args(&formatter.args)
+                .stdin(Stdio::piped())
+                .stdout(Stdio::piped())
+                .stderr(Stdio::piped());
+
+            let formatting_future = async move {
+                let mut process = process
+                    .spawn()
+                    .map_err(|e| FormatterError::SpawningFailed {
+                        command: formatter.command.clone(),
+                        error: e.kind(),
+                    })?;
+                {
+                    let mut stdin = process.stdin.take().ok_or(FormatterError::BrokenStdin)?;
+                    stdin
+                        .write_all(&text.bytes().collect::<Vec<u8>>())
+                        .await
+                        .map_err(|_| FormatterError::BrokenStdin)?;
```

Comment:
`.bytes().collect()` is very inefficient since it reallocates a new temporary vector to hold the entire file (which could be very large).

Let's copy the approach from #3180 and use `helix_view::document::to_writer` which will stream the contents:

```rust
helix_view::document::to_writer(&mut stdin, encoding::UTF_8, &text).await?;
```